Implement all JSON Patch operations

meppu opened the pull request:
Thank you for submitting this pull request! We appreciate you spending the time to work on these changes.

## What is the motivation?

Currently SurrealDB supports `add`, `remove`, and `replace` operations for JSON Patch. This pull request aims to implement the missing operations.

## What does this change do?

This PR introduces the `move`, `copy`, and `test` JSON Patch operations to SurrealDB.

It's important to note that there have been structural changes: the `Op` struct has been removed, and `Operation` is now an enum. Depending on usage, this change may be considered as a breaking change (though im not sure if these are exposed for public usage).

Additionally, I've added the `PatchTestFailure` error for test operations, intended for situations where a test operation fails for JSON Patch. I am open to any suggestions for refining the definition of this error.

## What is your testing strategy?

I added tests for new operations which is pretty much similar to existing ones.

## Is this related to any issues?

If this pull request is related to any other pull request or issue, or resolves any issues - then link all related pull requests and issues here.

## Have you read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)?

- [X] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

alyti commented:
One of the checks seems to be failing due to increased size of Error, it also looks like most Errors that deal with values have them as strings, might be worth stringifying the values in yours too.

meppu commented:
> One of the checks seems to be failing due to increased size of Error, it also looks like most Errors that deal with values have them as strings, might be worth stringifying the values in yours too.

Yeah i actually thinked about making them string, and also curious about current name too. `PatchTestFail` seems a bit weird to be honest. I would like to see other alternatives :)

meppu commented:
Renamed `PatchTestFail` error to `PatchTestFailure`, Updated error size for tests and some minor refactors.

I'm still not sure if these fields for error should be `Value` or `String`, since formatted message with `Value` pretty much works as expected.

tobiemh commented on lib/src/err/mod.rs:
```diff
@@ -107,6 +107,13 @@ pub enum Error {
 
+	/// Given test operation failed for JSON Patch
+	#[error("Given test operation failed for JSON Patch. Expected `{expected}`, but got `{got}` instead.")]
+	PatchTestFailure {
+		expected: Value,
+		got: Value,
+	},
```
Could we rename this to `PatchTest`? That will make it `Error::PatchTest` which removes the redundant 'Failure'!

meppu replied:
Indeed, I've been anticipating a review of this, as naming things isn't my strong suit 😅

tobiemh commented on lib/src/sql/value/value.rs:
```diff
@@ -2940,8 +2940,8 @@ mod tests {
 		assert_eq!(64, std::mem::size_of::<Value>());
-		assert_eq!(104, std::mem::size_of::<Error>());
-		assert_eq!(104, std::mem::size_of::<Result<Value, Error>>());
+		assert_eq!(136, std::mem::size_of::<Error>());
+		assert_eq!(136, std::mem::size_of::<Result<Value, Error>>());
```
It's important that the `Error` type is as small as possible for memory usage reasons when dealing with query responses. This is the reason that we convert most values to `String`. We actually want to reduce the size of `Error` below `104` rather than increase it. Therefore we would need to change the `Error::PatchTest` implementation so that it reduces the size. Then these line changes can be reverted.

meppu commented:
By the way, the WebSocket integration tests seem to have a problem with consistency. Sometimes they don't work, and other times they do...
